pandas/plotting/_matplotlib/hist.py

```diff
@@ -146,14 +146,23 @@ def _make_plot(self) -> None:
                 kwds["label"] = self.columns
                 kwds.pop("color")
 
-            y = reformat_hist_y_given_by(y, self.by)
-
             # We allow weights to be a multi-dimensional array, e.g. a (10, 2) array,
             # and each sub-array (10,) will be called in each iteration. If users only
             # provide 1D array, we assume the same weights is used for all iterations
             weights = kwds.get("weights", None)
-            if weights is not None and np.ndim(weights) != 1:
-                kwds["weights"] = weights[:, i]
+            if weights is not None:
+                if np.ndim(weights) != 1 and np.shape(weights)[-1] != 1:
+                    try:
+                        weights = weights[:, i]
+                    except IndexError:
+                        raise ValueError(
```

Comment:
Is there a tests that hits this exception?

Also we usually re-raise an exception from the original one (`raise ValueError(...) from err`)